File: __Viejo/Formularios/Agregar/frm_detalle_asoc_produccion.vb
Public Class frm_detalle_asoc_produccion
    Dim id_item As Integer
    Dim id_produccion As Integer
    Dim dt_cantidades_items_asociados As DataTable

    Public Sub New()

        ' Esta llamada es exigida por el diseñador.
        InitializeComponent()

        ' Agregue cualquier inicialización después de la llamada a InitializeComponent().

    End Sub

    Public Sub New(ByVal _id_item As Integer, Optional ByVal _id_produccion As Integer = -1)

        ' Esta llamada es exigida por el diseñador.
        InitializeComponent()

        ' Agregue cualquier inicialización después de la llamada a InitializeComponent().
        id_item = _id_item
        id_produccion = _id_produccion
    End Sub
    Private Sub frm_detalle_asoc_produccion_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        If id_produccion <> -1 Then
            dt_cantidades_items_asociados = Traer_Cantidades_Enviadas_Items_Asociados_Default_Produccion(id_item, id_produccion)
            dg_view.ReadOnly = True
            cmd_ok.Enabled = False
        Else
            dt_cantidades_items_asociados = Traer_Cantidades_Enviadas_Items_Asociados_Default_Produccion(id_item)
        End If

        dg_view.DataSource = dt_cantidades_items_asociados
    End Sub

    Private Sub cmd_ok_Click(sender As Object, e As EventArgs) Handles cmd_ok.Click
        For Each row As DataRow In dt_cantidades_items_asociados.Rows
            addItemAsocProducciontmp(row.Item(4), row.Item(5), row.Item(6), row.Item(3))
        Next

        closeandupdate(Me)
    End Sub
End Class

File: Formularios/Agregar/frm_detalle_asoc_produccion.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace Centrex
{
    public partial class frm_detalle_asoc_produccion
    {
        private int id_item;
        private int id_produccion;
        private DataTable dt_cantidades_items_asociados = new DataTable();

        public frm_detalle_asoc_produccion()
        {

            // Esta llamada es exigida por el diseñador.
            InitializeComponent();

            // Agregue cualquier inicialización después de la llamada a InitializeComponent().

        }

        public frm_detalle_asoc_produccion(int _id_item, int _id_produccion = -1)
        {

            // Esta llamada es exigida por el diseñador.
            InitializeComponent();

            // Agregue cualquier inicialización después de la llamada a InitializeComponent().
            id_item = _id_item;
            id_produccion = _id_produccion;
        }
        private void frm_detalle_asoc_produccion_Load(object sender, EventArgs e)
        {
            var resultado = (id_produccion != -1)
                ? asocitems.Traer_Cantidades_Enviadas_Items_Asociados_Default_Produccion(id_item.ToString(), id_produccion)
                : asocitems.Traer_Cantidades_Enviadas_Items_Asociados_Default_Produccion(id_item.ToString());

            // 🔹 Usa tu ToDataTable<T> genérico con el tipo real del resultado
            var dataList = resultado?.DataMaterializada as IEnumerable<object>;
            DataTable dt = new DataTable();

            if (dataList != null && dataList.Any())
            {
                // Tomamos el tipo del primer elemento para inferir T
                var firstType = dataList.First().GetType();
                var genericMethod = typeof(generales).GetMethod("ToDataTable", new[] { typeof(IEnumerable<>).MakeGenericType(firstType) });
                if (genericMethod != null)
                    dt = (DataTable)genericMethod.Invoke(null, new object[] { dataList });
            }

            dg_view.DataSource = dt;

            if (id_produccion != -1)
            {
                dg_view.ReadOnly = true;
                cmd_ok.Enabled = false;
            }
        }


        private void cmd_ok_Click(object sender, EventArgs e)
        {
            foreach (DataRow row in dt_cantidades_items_asociados.Rows)
                producciones.addItemAsocProducciontmp(Conversions.ToInteger(row[4]), Conversions.ToInteger(row[5]), Conversions.ToInteger(row[6]), Conversions.ToInteger(row[3]));

            closeandupdate(this);
        }
    }
}
